### PR TITLE
volume-basic-test.sh: switch to .tplg parsing to test topologies v2

### PR DESCRIPTION
**Pull request by marc-hb**
4 commits

Until now this test was "screen-scraping" amixer output. This was not
compatible with v2 topologies which means the test was always SKIP for
everything developped on the main branch. Switch to parsing topologies
to extract volume kcontrols.

Fixes: https://github.com/thesofproject/sof-test/issues/1069, see also discussion in earlier attempt https://github.com/thesofproject/sof-test/pull/1068.

### CODE REVIEW

**Comment by marc-hb** [comment]
In some test configurations there are many more "volume" kcontrols than found in the topology. Is this expected?

Probably the most extreme example is on stable-v2.2 / topology v1:
```
jf-cml-hel-rt5682-05:~$ amixer | grep -B1 Capabilities:.*volume
```
   38 "volume" controls!! for instance:
```
numid=2,iface=MIXER,name='CBJ Boost Volume'
numid=1,iface=MIXER,name='DAC1 Playback Volume'
...
```
The vast majority of these are NOT in `sof-tplg/sof-cml-rt1011-rt5682.tplg` which has only PGA1 to PGA7.

Similarly,
https://sof-ci.01.org/softestpr/PR1193/build420/devicetest/index.html?model=LNLM_SDW_AIOC&testcase=volume-basic-test

```
ba-lnlm-rvp-sdw-03:~$ amixer | grep -B1 Capabilities:.*volume
```

  23 volume controls, among others:

```  
'Pre Mixer Deepbuffer Jack Out', 'rt1316-1 Left I Tag Select', 'rt711 FU15 Gain', 'rt714 FU0C Boost', etc.
```
Yet `sof-ipc4-tplg/sof-lnl-rt711-l0-rt1316-l23-rt714-l1.tplg` has only 5 PGAs `gain.N.M`:

```
PGA: gain.0.1
Pre Mixer Jack Out Playback Volume
PGA: gain.1.1
Post Mixer Jack Out Playback Volume
PGA: gain.15.1
Pre Mixer Deepbuffer Jack Out Volume
PGA: gain.20.1
Amplifier Volume
PGA: gain.21.1
Main Amplifier Volume
```

**Comment by plbossart** [comment]
````
numid=2,iface=MIXER,name='CBJ Boost Volume'
numid=1,iface=MIXER,name='DAC1 Playback Volume'
````

Those are codec kcontrols.

**Comment by marc-hb** [comment]
> Those are codec kcontrols.

Thanks, @ranj063 just told me: no need to test these (at least not in this "basic" test)

File: tools/topo2_list_vol_kcontrols.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                kc
+                for kc in gb.kcontrols
+                if kc.hdr.type == DapmType.MIXER.name and kc.body.max != 1
+            ]
+
+            assert len(volume_kcontrols) == 1
+
+            # TODO: find how the kernel makes this "topology v1 versus v2"
+            # decision and mimic it here.
+            pga_prefix = (
```
Comment (marc-hb) [inline tools/topo2_list_vol_kcontrols.py:51]: Clue from @ranj063 (thx!) `no_wname_in_kcontrol_name`

https://github.com/thesofproject/sof/blob/fdb67ee782808f6d5488a53441cdecd75f458677/tools/topology/topology2/include/components/widget-common.conf#L121

**Review by marc-hb — COMMENTED** [review]
New volume tests passing in:
- LNL https://sof-ci.01.org/softestpr/PR1193/build426/devicetest/index.html
- MTL https://sof-ci.01.org/softestpr/PR1193/build428/devicetest/index.html
- TGL IPC4/topoV2 https://sof-ci.01.org/softestpr/PR1193/build427/devicetest/index.html

Old tests still passing "the new way" on stable-v2.2
- https://sof-ci.01.org/sofpr/PR9113/build4600/devicetest/index.html

All test results looking good!

**Comment by marc-hb** [comment]
> Those are codec kcontrols.

Wait, is `numid=9,iface=MIXER,name='Master Playback Volume'` really a codec kcontrol? The name really does not sound like it (and it's not in `sof-ipc4-tplg/sof-hda-generic-ace1-4ch.tplg`)

https://sof-ci.01.org/softestpr/PR1193/build420/devicetest/index.html?model=LNLM_RVP_HDA&testcase=volume-basic-test

**Comment by plbossart** [comment]
You could skip the SOF driver with 'options snd-intel-dspcfg dsp_driver=1' and you would see which controls are from the codec. 'Master Playback' is rather typical of HDaudio codecs.

That said, we should indeed have added an SOF prefix to make all firmware-controls self-explanatory.

**Comment by marc-hb** [comment]
@ranj063 , @plbossart , I flipped the NO_WNAME setting and so one .tplg file hits the 44 characters limit in order to test the kcontrol truncation issue spotted in:
- https://github.com/thesofproject/sof-test/pull/1068#issuecomment-1608537201

```
ubuntu@jf-adlp-rvp-nocodec-8:~/$ amixer controls

numid=14,iface=MIXER,name='gain.17.1 Post Demux Port0 2nd Capture Volu'
numid=15,iface=MIXER,name='gain.18.1 Post Demux DMIC SFX1 Capture Volu'
```



Then I ran the new test in this PR. It passes without any issue because either `amixer` or the kernel truncates BOTH on output and on input. In other words, this command works just fine:

```
ubuntu@jf-adlp-rvp-nocodec-8:~$ amixer cget name='gain.17.1 Post Demux Port0 2nd Capture VolumeTRUNCATED_DOES_NOT_MATTER'
numid=14,iface=MIXER,name='gain.17.1 Post Demux Port0 2nd Capture Volu'
  ; type=INTEGER,access=rw---R--,values=2,min=0,max=45,step=0
  : values=45,45
  | dBscale-min=-90.00dB,step=2.00dB,mute=1
```

Note the name is NOT truncated in the .tplg file.

**Comment by marc-hb** [comment]
Tests results look good, this is ready.

File: tools/topo_vol_kcontrols.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    # Either 1 volume kcontrol, or 1 volume + 1 switch
+    assert gain_block.widget.num_kcontrols in (1, 2)
+
+    # A switch is either a DapmType.SWITCH, or DapmType.MIXER
+    # with a max "volume" = 1. Don't include switches here.
+    volume_kcontrols = [
+        kc
+        for kc in gain_block.kcontrols
+        if kc.hdr.type == DapmType.MIXER.name and kc.body.max != 1
```
Comment (plbossart) [inline tools/topo_vol_kcontrols.py:48]: indentation doesn't look like usual python, is this intentional?

File: tools/topo_vol_kcontrols.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# This could probably be moved to tplgtool2.py?
+def has_wname_prefix(widget):
+    """Is the kcontrol name prefixed with the widget name? ("PGAxx" or "Dmicxx")
+    Check SOF_TKN_COMP_NO_WNAME_IN_KCONTROL_NAME"""
+
+    wname_elems = [
+        prv.elems
+        for prv in widget.priv
+        if prv.elems[0].token
+        == SofVendorToken.SOF_TKN_COMP_NO_WNAME_IN_KCONTROL_NAME.name
```
Comment (plbossart) [inline tools/topo_vol_kcontrols.py:70]: same here, indentation is probably something intentional but someone like me will never figure out what this does.

File: tools/tplgtool2.py [hunk 4]
```diff
@@ -1330,6 +1330,7 @@ def main():
         errors = 0
         for f in files:
             tplg = GroupedTplg(tplgFormat.parse_file(f))
+            assert set(dump_types) <= set(supported_dump), f"unsupported type in {dump_types}"
```
Comment (plbossart) [inline tools/tplgtool2.py:1334]: no idea what this does.

**Review by plbossart — COMMENTED** [review]
way beyond my comfort zone....

**Comment by plbossart** [comment]
jf-lnlm-rvp-sdw-1 seems to be in bad shape @fredoh9, was the BIOS revert applied?

````
[  233.238872] kernel: soundwire sdw-master-0-2: Controller Timed out on bank switch
[  233.238996] kernel: soundwire sdw-master-0-2: multi link bank switch failed: -110
````

File: tools/topo_vol_kcontrols.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            # debug
+            # print(f"{gb.widget.id}: {gb.widget.name}")
+            print_volume_kcontrols(gb)
+
+
+def print_volume_kcontrols(gain_block):
+    "Print volume kcontrols"
+
+    # Either 1 volume kcontrol, or 1 volume + 1 switch
+    assert gain_block.widget.num_kcontrols in (1, 2)
```
Comment (ranj063) [inline tools/topo_vol_kcontrols.py:41]: can we print an error message here?

File: tools/topo_vol_kcontrols.py (hunk 2 above)
Comment (ranj063) [reply tools/topo_vol_kcontrols.py:48]: python list comprehension @plbossart

**Comment by fredoh9** [comment]
> jf-lnlm-rvp-sdw-1 seems to be in bad shape @fredoh9, was the BIOS revert applied?
> 
> ```
> [  233.238872] kernel: soundwire sdw-master-0-2: Controller Timed out on bank switch
> [  233.238996] kernel: soundwire sdw-master-0-2: multi link bank switch failed: -110
> ```

I did and run-all-tests.sh tested. I will move it to maintenance mode and double check again.

File: tools/topo_vol_kcontrols.py (hunk 5 above)
Comment (marc-hb) [reply tools/topo_vol_kcontrols.py:41]: `assert` already prints the entire line + a stack trace. Do you want anything extra?

AFAIK this is never supposed to happened, or is it?

File: tools/topo_vol_kcontrols.py (hunk 2 above)
Comment (marc-hb) [reply tools/topo_vol_kcontrols.py:48]: > python list comprehension 

Yes.

This new script was formatted with `black`; recommended. Warning: formats _in place_, make a backup first if needed.

File: tools/topo_vol_kcontrols.py (hunk 3 above)
Comment (marc-hb) [reply tools/topo_vol_kcontrols.py:70]: List comprehension here too. This is just building a list from another, filtered list.

File: tools/tplgtool2.py (hunk 4 above)
Comment (marc-hb) [reply tools/tplgtool2.py:1334]: Before this commit, a typo in the command-line argument was silently producing no output.

**Comment by marc-hb** [comment]
SOFCI TEST

**Review by marc-hb — COMMENTED** [review]
> way beyond my comfort zone....

Thanks for taking a look anyway! In theory, you shouldn't need to be proficient in Python to at least follow most of what is going on. If you cannot, then the functions and variable names are probably not good enough.


> jf-lnlm-rvp-sdw-1 seems to be in bad shape @fredoh9, was the BIOS revert applied?

https://sof-ci.01.org/softestpr/PR1193/build431/devicetest/index.html was an OLD run!

I'm re-running tests  now (I was waiting for daily runs to finish)

File: tools/topo_vol_kcontrols.py (hunk 5 above)
Comment (ranj063) [reply tools/topo_vol_kcontrols.py:41]: ok great. nothing more needed. i was worried assert would silently fail

**Review by ranj063 — APPROVED** [review]

**Pull request merged**